python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,292 @@
+}
+
+template <typename T>
+__device__ double BlockStd(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ double var;
+  if (block.thread_rank() == 0) { var = 0; }
+  block.sync();
+
+  device_var<T>(block, data, size, &var);
+  return sqrt(var);
+}
+
+template <typename T>
+__device__ double BlockVar(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ double block_var;
+  if (block.thread_rank() == 0) { block_var = 0; }
+  block.sync();
+
+  device_var<T>(block, data, size, &block_var);
+  return block_var;
+}
+
+template <typename T>
+__device__ T BlockMax(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  auto local_max = []() {
+    if constexpr (std::is_floating_point_v<T>) { return -std::numeric_limits<T>::max(); }
+    return std::numeric_limits<T>::min();
+  }();
+  __shared__ T block_max;
+  if (block.thread_rank() == 0) { block_max = local_max; }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    local_max = max(local_max, data[idx]);
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{block_max};
+  ref.fetch_max(local_max, cuda::std::memory_order_relaxed);
+
+  block.sync();
+
+  return block_max;
+}
+
+template <typename T>
+__device__ T BlockMin(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  auto local_min = std::numeric_limits<T>::max();
+  __shared__ T block_min;
+  if (block.thread_rank() == 0) { block_min = local_min; }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    local_min = min(local_min, data[idx]);
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{block_min};
+  ref.fetch_min(local_min, cuda::std::memory_order_relaxed);
+
+  block.sync();
+
+  return block_min;
+}
+
+template <typename T>
+__device__ int64_t BlockIdxMax(T const* data, int64_t* index, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_max;
+  __shared__ int64_t block_idx_max;
+
+  auto local_max = []() {
+    if constexpr (std::is_floating_point_v<T>) { return -std::numeric_limits<T>::max(); }
+    return std::numeric_limits<T>::min();
+  }();
+  auto local_idx_max = std::numeric_limits<int64_t>::max();
+
+  if (block.thread_rank() == 0) {
+    block_max     = local_max;
+    block_idx_max = local_idx_max;
+  }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto const current_data = data[idx];
+    if (current_data > local_max) {
+      local_max     = current_data;
+      local_idx_max = index[idx];
+    }
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{block_max};
+  ref.fetch_max(local_max, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  if (local_max == block_max) {
```

Comment:
Initially I was hesitant about the nuances of this equality check in the case of `NaN` values. Then I got to thinking a little more: 

I think we need to carefully test the behavior of this `BlockIdxMax` code if we receive a group with all `NaN` entries and/or all `+Inf` entries. It might return `std::numeric_limits<int64_t>::max()` as the index, which is probably illegal.

You might need a variable that tracks whether the `current_data > local_max` was ever reached, and return `0` if that's never reached?

Reply:
Any other corner cases you can think of other than all `NaN`, all `+Inf` or all `-Inf`?